dropshot/src/api_description.rs

```diff
@@ -914,11 +931,15 @@ impl<Context: ServerContext> ApiDescription<Context> {
                     openapiv3::StatusCode::Code(code.as_u16()),
                     openapiv3::ReferenceOr::Item(response),
                 );
+                let ApiEndpointErrorMetadata { ref name, ref schema } =
+                    endpoint.error_type;
+                let err_ref = openapiv3::ReferenceOr::Reference {
+                    reference: format!("#/components/responses/{name}"),
+                };
+                errors
+                    .entry(endpoint.error_type.name.clone())
+                    .or_insert_with(|| schema(&mut generator));
```

Comment:
We stuff the schemas for each endpoint's error type into a big map, so that we can put the schemas in `components.responses` in order to allow endpoints that return the same Rust error type to also return the same type in the OpenAPI document.

Unfortunately, though, I think just keying this by the name of the schema is probably not sufficient, because I believe `schemars`' `derive(JsonSchema)` just uses the type's name, _not_ its fully qualified name. Which, I think, means that if you have an endpoint that returns a `foo::bar::Error` and another endpoint that returns a `baz::quux::Error`, and these are different Rust types, their schemas will clobber each other. The key for the map should probably be the fully-qualified type name including the module path, and we might need to massage that into something we can use as the reference in the OpenAPI doc.

I think this needs to be fixed before this PR is mergeable; I'm leaving this comment to get it written down while I remember.

Reply:
Went and implemented this in e9728118742f6891bf0c5f427fc93bcd471c8003 --- it's a bit hairy because I wanted to generate reasonably nice unambiguous names for the error responses. It turns out that `schemars` will internally disambiguate _schema names_ by appending a number to them, which I had missed, but it's still necessary to do our own disambiguation as well if we want to reference schemas in `components.schemas` from `components.responses`, since we still have to generate the references to them ourselves. 

It would certainly be nicer if there was a way to get `schemars` to use my prettier unambiguous names for the schemas themselves rather than `Error1`, `Error2`,  but I think this is good enough:

```json
{
  "openapi": "3.0.3",
  "info": {
    "title": "Custom Error Example",
    "version": "1.0"
  },
  "paths": {
    "/pdflatex/{filename}": {
      "get": {
        "operationId": "get_pdflatex",
        "parameters": [
          {
            "in": "path",
            "name": "filename",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ],
        "responses": {
          "201": {
            "description": "successful creation",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Pdf"
                }
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/LatexError"
          },
          "5XX": {
            "$ref": "#/components/responses/LatexError"
          }
        }
      }
    },
    "/ye-flask": {
      "get": {
        "summary": "Gets ye flask.",
        "operationId": "get_ye_flask",
        "responses": {
          "200": {
            "description": "successful operation",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Flask"
                }
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/FlaskError"
          },
          "5XX": {
            "$ref": "#/components/responses/FlaskError"
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "Error": {
        "description": "Errors returned by pdflatex.\n\nGood luck figuring out what these mean!",
        "oneOf": [
          {
            "description": "An hbox is overfull.",
            "type": "object",
            "properties": {
              "OverfullHbox": {
                "type": "object",
                "properties": {
                  "badness": {
                    "description": "The amount of badness in the overfull hbox.",
                    "type": "integer",
                    "format": "uint",
                    "minimum": 0
                  }
                },
                "required": [
                  "badness"
                ]
              }
            },
            "required": [
              "OverfullHbox"
            ],
            "additionalProperties": false
          },
          {
            "description": "Like an overfull hbox, except the opposite of that.",
            "type": "object",
            "properties": {
              "UnderfullHbox": {
                "type": "object",
                "properties": {
                  "badness": {
                    "description": "This one also has badness.",
                    "type": "integer",
                    "format": "uint",
                    "minimum": 0
                  }
                },
                "required": [
                  "badness"
                ]
              }
            },
            "required": [
              "UnderfullHbox"
            ],
            "additionalProperties": false
          }
        ]
      },
      "Error2": {
        "type": "object",
        "properties": {
          "reason": {
            "type": "string"
          }
        },
        "required": [
          "reason"
        ]
      },
      "Flask": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string"
          }
        },
        "required": [
          "name"
        ]
      },
      "Pdf": {
        "type": "object",
        "properties": {
          "filename": {
            "type": "string"
          }
        },
        "required": [
          "filename"
        ]
      }
    },
    "responses": {
      "LatexError": {
        "description": "Error",
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error"
            }
          }
        }
      },
      "FlaskError": {
        "description": "Error",
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/Error2"
            }
          }
        }
      }
    }
  }
}
```